### ISSUE TITLE
Applications are not checked for validity on save

### ISSUE THREAD

**Issue by JulianFeinauer** [drogue-iot/drogue-cloud#264]
It seems that you can currently enter an invalid Application definition (as yaml) and thats get stored without an Issue.

Expected Behavior:
This should produce an error internaly and return an error code and not get saved

**Comment by ctron** [comment]
True. Currently that is "on purpose". As we only enforce the metadata section and the fact that there are a `.status` and `.spec` field of the type `map`.

As discussed in the chat, it makes of course sense to validate (before storing). Some of the content however cannot be validated by Drogue Cloud, as it doesn't know about the content. A possible solution would be to configure (opt-in) web hook validation on the application level, which would be able to validate.

So the logic could be:

* Run internal validation
* Contact all configured webhooks for validation (even in parallel)

Any rejection would reject the whole operations. Maybe a webhook could also alter the content (which might make it tricky to parallelize).

I think it would make sense to use Knative serving style endpoints again (as we do with payload validation/enrichment).

PR's welcome :) … I would also say that a PR could start with some internal validation first. Assuming this is wrapped by an `async` operation (which can later be extended easily with webhooks).

**Comment by lulf** [comment]
@JulianFeinauer At the moment, it is possible to add custom fields within .spec. and .status for the Application and Device types. 

One such example application is https://github.com/drogue-iot/drogue-ajour, which extends them with additional information related to firmware updates, and can be deployed by the 'user' of Drogue Cloud.

**Comment by ctron** [comment]
The service responsible for writing the application to the persistent storage is the `device-management-service`. That contains two layers (Rust modules): `endpoints` which defines the REST endpoints and forward the requests to some persistence service, and `service` which defines the persistence services and their implementations (currently postgres only).

I guess it doesn't make sense to start any database operation before validating the data. So I would assume the validation part triggered by the `endpoints` layer before executing calls on the `service` layer.

**Comment by JulianFeinauer** [comment]
yes that makes sense to put it in `endpoints`, indeed.
After @lulf response I realized that there may be different kinds of validations.
The validation I was looking for, was a situation where there is a `Dialog` (from Drogue Core) which will later be serialized from the `application`. But some kind of issues (in my case there was an ENUM and i used a string not in the enum) this leads to the `dialog` not being parseable and thus silent failure.

Which is a totally different kind of validation then to check for custom fields.

So we should also perhaps consider how to do that.

**Comment by ctron** [comment]
A dialog?

**Comment by JulianFeinauer** [comment]
Sorry, a dialect, i mean.
